node/src/storage.rs

```diff
@@ -0,0 +1,165 @@
+    node_data: Option<PersistentNodeData>,
+}
+
+#[async_trait]
+impl SecretNodeStorage for MemoryNodeStorage {
+    async fn store(&mut self, data: &PersistentNodeData) -> Result<()> {
+        self.node_data = Some(data.clone());
+        Ok(())
+    }
+
+    async fn load(&self) -> Result<Option<PersistentNodeData>> {
+        Ok(self.node_data.clone())
+    }
+}
+
+struct SecretManagerNodeStorage {
+    secret_manager: SecretManager<HttpsConnector<HttpConnector>>,
+    gcp_project_id: String,
+    sk_share_secret_id: String,
+}
+
+impl SecretManagerNodeStorage {
+    async fn new(gcp_project_id: String, sk_share_secret_id: String) -> Result<Self> {
+        let client = hyper::Client::builder().build(
+            hyper_rustls::HttpsConnectorBuilder::new()
+                .with_native_roots()
+                .https_or_http()
+                .enable_http1()
+                .enable_http2()
+                .build(),
+        );
+        let opts = ApplicationDefaultCredentialsFlowOpts::default();
+        let authenticator = match ApplicationDefaultCredentialsAuthenticator::builder(opts).await {
+            ApplicationDefaultCredentialsTypes::InstanceMetadata(auth) => auth.build().await?,
+            ApplicationDefaultCredentialsTypes::ServiceAccount(auth) => auth.build().await?,
+        };
+        let secret_manager = SecretManager::new(client.clone(), authenticator.clone());
+        Ok(Self {
+            secret_manager,
+            gcp_project_id,
+            sk_share_secret_id,
+        })
+    }
+}
+
+#[async_trait]
+impl SecretNodeStorage for SecretManagerNodeStorage {
+    async fn store(&mut self, data: &PersistentNodeData) -> Result<()> {
+        self.secret_manager
+            .projects()
+            .secrets_add_version(
+                AddSecretVersionRequest {
+                    payload: Some(SecretPayload {
+                        data: Some(serde_json::to_vec(data)?),
+                        ..Default::default()
+                    }),
+                },
+                &format!(
+                    "projects/{}/secrets/{}",
+                    self.gcp_project_id, self.sk_share_secret_id
+                ),
+            )
+            .doit()
+            .await?;
+        Ok(())
+    }
+
+    async fn load(&self) -> Result<Option<PersistentNodeData>> {
+        let (_, response) = self
+            .secret_manager
+            .projects()
+            .secrets_versions_access(&format!(
+                "projects/{}/secrets/{}/versions/latest",
+                self.gcp_project_id, self.sk_share_secret_id
+            ))
+            .doit()
+            .await?;
+        match response.payload {
+            // GCP does not allow to upload empty secrets, so we reserve 1-byte values as a
+            // placeholder for empty secrets.
+            Some(SecretPayload {
+                data: Some(data), ..
+            }) if data.len() > 1 => Ok(Some(serde_json::from_slice(&data)?)),
+            _ => {
```

Comment:
if it's missing and return Ok(None), seems like the node will be started with None as the PersistentNodeData. Is that acceptable?

Reply:
Yeah, it just means we don't have any key share saved and we presume the node just doesn't have one yet (has to either generate one in collaboration with other participants, or try to join an existing set via resharing).